DemoViper/Sources/Extension/Decodable+Initialize.swift

```diff
@@ -0,0 +1,19 @@
+//
+//  Decodable+Initialize.swift
+//  DemoViper
+//
+//  Created by Nguyen M. Tam on 4/20/20.
+//  Copyright © 2020 Nimble Co. Ltd. All rights reserved.
+//
+
+import Foundation
+
+extension Decodable {
+
+    init?(from json: String) {
+        guard
+            let data = json.data(using: .utf8),
+            let value = try? JSONDecoder().decode(Self.self, from: data) else { return nil }
```

Comment:
For shorter
```swift
guard let data = json.data(using: .utf8) else { return nil }
self = try? JSONDecoder().decode(Self.self, from: data)
```

Reply:
Not the right way to implement a failable initializer. Please refer to this document: https://docs.swift.org/swift-book/LanguageGuide/Initialization.html
A failable initializer creates an optional value of the type it initializes. We write `return nil` within a failable initializer to indicate a point at which initialization failure can be triggered. Strictly speaking, initializers do not return a value. Rather, their role is to ensure that self is **fully and correctly initialized** by the time that initialization ends.
Which means we can assigne `self.someVariable` with `try? JSONDecoder().decode(Self.self, from: data)` but cannot do that with `self`.